r1cs/gadgets/std/src/fields/nonnative/reduce.rs

```diff
@@ -98,110 +76,294 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
+    /// `` 
+    /// Implements the grouping technique from [[Kosba et al]] to reduce the number of 
+    /// constraints for the carries. 
+    /// Assumes that
+    /// ``
+    ///     bits_per_limb + surfeit + 2 <= ConstraintF::CAPACITY,
+    /// ``
+    /// and `shift_per_limb >= 2`. 
+    /// 
+    /// [Kosba et al]: https://ieeexplore.ieee.org/document/8418647
+    /// 
+    // Costs
+    // ``
+    //      (num_groups - 1) * (1 + bits_per_limb + surfeit + 2 - shift_per_limb) + 2
+    // ``
+    // constraints, where `1 <= num_groups <= num_limbs` is the number of groups, determined
+    // by `num_groups = Ceil[num_limbs / S]` with 
+    // ``
+    //  S - 1 = Floor[
+    //          (ConstraintF::CAPACITY - 2 - (bits_per_limb + surfeit)) / shift_per_limb
+    //      ].
+    // ``
+    // TODO: can be slightly optimized by using the number of additions over normal
+    // form instead of `surfeit`.
     pub fn group_and_check_equality<CS: ConstraintSystemAbstract<ConstraintF>>(
         mut cs: CS,
+        // The additional number of bits beyond `bits_per_limb`. Hence the current
+        // limbsize is `bits_per_limb + surfeit`.
         surfeit: usize,
+        // The number of bits  
         bits_per_limb: usize,
+        // defines arity of the limb representation, i.e. `A= 2^shift_per_limb`.
+        // MUST be >= 2.
         shift_per_limb: usize,
         left: &[FpGadget<ConstraintF>],
         right: &[FpGadget<ConstraintF>],
     ) -> Result<(), SynthesisError> {
+        
         let zero = FpGadget::<ConstraintF>::zero(cs.ns(|| "hardcode zero"))?;
 
         let mut limb_pairs = Vec::<(FpGadget<ConstraintF>, FpGadget<ConstraintF>)>::new();
-        let num_limb_in_a_group = (ConstraintF::size_in_bits()
-            - 1
+        // We group the limbs so that we can handle their sum
+        // ``
+        //      group_total = limb[0] * 1 + limb[1] * A + ... + limb[S-1] * A^{S-1},
+        // ``
+        // within a single constraint field element. Assuming that `A >= 2` we have 
+        // `2 * A^i <= A^{i+1}` and therefore
+        // `` 
+        //   limb[0] + limb[1] * A < 2 * A * 2^limb_size
+        //   limb[0] + limb[1] * A + limb[2] * A^2 < 2 * A^2 * 2^limb_size
+        //   ...
+        //   limb[0] + limb[1] * A + ... + limb[S-1] * A^{S-1} <
+        //                       < 2 * A^{S-1} * 2^limb_size.
+        // ``
+        // Hence `len(group_total) <= bits_per_group`, with 
+        // ``
+        //     bits_per_group = limb_size + 1 + (S - 1) * shift_per_limb.
+        // ``
+        // To assure that the following operations on the totals do not exceed the capacity, 
+        // it is sufficient to demand 
+        // ``
+        //      bits_per_group + 1 <= ConstraintF::CAPACITY,
+        // ``
+        // as described in detail below. This yields the security condition
+        // ``
+        //      bits_per_limb + surfeit + (S - 1) * shift_per_limb + 2 <= ConstraintF::CAPACITY,
+        // ``
```

Comment:
I would specify in the comment that the value below is the maximum S which achieves the inequality above.